source/fuzz/transformation_move_instruction.cpp

```diff
@@ -0,0 +1,430 @@
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "source/fuzz/transformation_move_instruction.h"
+
+#include <queue>
+#include <unordered_set>
+#include <vector>
+
+#include "source/fuzz/fuzzer_util.h"
+#include "source/fuzz/instruction_descriptor.h"
+
+namespace spvtools {
+namespace fuzz {
+
+TransformationMoveInstruction::TransformationMoveInstruction(
+    const protobufs::TransformationMoveInstruction& message)
+    : message_(message) {}
+
+TransformationMoveInstruction::TransformationMoveInstruction(
+    const protobufs::InstructionDescriptor& insert_before,
+    const protobufs::InstructionDescriptor& target) {
+  *message_.mutable_insert_before() = insert_before;
+  *message_.mutable_target() = target;
+}
+
+bool TransformationMoveInstruction::IsApplicable(
+    opt::IRContext* ir_context, const TransformationContext& /*unused*/) const {
+  // |insert_before| must be valid.
+  auto* insert_before_inst =
+      FindInstruction(message_.insert_before(), ir_context);
+  if (!insert_before_inst) {
+    return false;
+  }
+
+  auto* target_inst = FindInstruction(message_.target(), ir_context);
+  if (!target_inst) {
+    return false;
+  }
+
+  return CanMoveInstruction(
+      ir_context,
+      fuzzerutil::GetIteratorForInstruction(
+          ir_context->get_instr_block(insert_before_inst), insert_before_inst),
+      fuzzerutil::GetIteratorForInstruction(
+          ir_context->get_instr_block(target_inst), target_inst));
+}
+
+void TransformationMoveInstruction::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto* insert_before_inst =
+      FindInstruction(message_.insert_before(), ir_context);
+  assert(insert_before_inst && "|insert_before| descriptor is invalid");
+
+  auto* target_inst = FindInstruction(message_.target(), ir_context);
+  assert(target_inst && "|target| descriptor is invalid");
+
+  assert(insert_before_inst != target_inst &&
+         "Can't insert target instruction before itself");
+
+  // Target instruction is not disposed after removal.
+  target_inst->RemoveFromList();
+
+  insert_before_inst->InsertBefore(
+      std::unique_ptr<opt::Instruction>(target_inst));
+
+  ir_context->set_instr_block(target_inst,
+                              ir_context->get_instr_block(insert_before_inst));
+
+  ir_context->InvalidateAnalysesExceptFor(opt::IRContext::kAnalysisNone);
+}
+
+protobufs::Transformation TransformationMoveInstruction::ToMessage() const {
+  protobufs::Transformation result;
+  *result.mutable_move_instruction() = message_;
+  return result;
+}
+
+bool TransformationMoveInstruction::IsMemoryBarrier(SpvOp opcode) {
```

Comment:
IsMemoryBarrier isn't the right name here, because CopyMemory and Store are not memory barriers.

I think "ModifiesOrOrdersMemory" would be a better name.  This captures Store, CopyMemory, etc., which modify memory, and memory barrier instructions, which order memory.

Reply:
Fixed in 587c227.